src/condvar.rs

```diff
@@ -572,6 +714,113 @@ mod tests {
         drop(g);
     }
 
+    fn spawn_wait_while_notifier(
+        mutex: Arc<Mutex<u32>>,
+        cv: Arc<Condvar>,
+        num_iters: u32,
+        timeout: Option<Instant>,
+    ) -> JoinHandle<()> {
+        thread::spawn(move || {
+            for epoch in 1..=num_iters {
+                // spin to wait for main test thread to block
+                // before notifying it to wake back up and check
+                // its condition.
+                let mut sleep_backoff = Duration::from_millis(1);
+                let _mutex_guard = loop {
+                    let mutex_guard = mutex.lock();
+
+                    if let Some(timeout) = timeout {
+                        if Instant::now() >= timeout {
+                            return;
+                        }
+                    }
+
+                    if *mutex_guard == epoch {
+                        break mutex_guard;
+                    }
+
+                    drop(mutex_guard);
+
+                    // give main test thread a good chance to
+                    // acquire the lock before this thread does.
+                    sleep(sleep_backoff);
+                    sleep_backoff *= 2;
+                };
+
+                cv.notify_one();
+            }
+        })
+    }
+
+    #[test]
+    fn wait_while_until_internal_does_not_wait_if_initially_false() {
+        let mutex = Arc::new(Mutex::new(()));
+        let cv = Arc::new(Condvar::new());
+
+        let mut mutex_guard = mutex.lock();
+
+        let result = cv
+            .wait_while_until_internal(&mut mutex_guard, |_| false, None)
+            .0;
+
+        assert!(result.num_iters() == 0);
+    }
+
+    #[test]
+    fn wait_while_until_internal_times_out_before_false() {
+        let mutex = Arc::new(Mutex::new(0));
+        let cv = Arc::new(Condvar::new());
+
+        let condition = |counter: &mut u32| {
+            *counter += 1;
+            true
+        };
+
+        let mut mutex_guard = mutex.lock();
+        let timeout = Some(Instant::now() + Duration::from_millis(50));
```

Comment:
This timeout may result in a flaky test if it's too aggressive, but in practice, I think 50ms should be fine and it doesn't meaningfully impact test execution speed. It would be great if we could inject a virtual clock and then manually advance the time deterministically, but that would require quite a bit of work + complicate the API for questionable benefits.

Reply:
I'd bump the timeout to 500ms just to be safe. The difference in CI times is imperceptible, and tests run in parallel anyways.